### ISSUE TITLE
Custom object attributes in ipelets

### ISSUE THREAD

**Issue by thobl** [otfried/ipe#409]
I would love to have a feature that allows me to write `obj:set("myCustomProperty", "value")` and then includes this property in the xml of the object.

The specific application I currently have for this feature is this [Manim ipelet](https://github.com/thobl/ipelets/tree/master/manim), which converts Ipe animations to smooth video animations via the python library Manim. For this, the user needs the option to identify different objects on different views with each other, so that the ipelet knows which objects should be transformed into each other.

Right now, my solution is quite hacky: I abuse the opacity attribute of the objects to store labels for them. Of course this only works as long as the user does not want to use transparency. Beyond labels, such custom attributes would also make it possible to further customize the animations (e.g., let the user choose which objects should fade out and which should just disappear).

**Comment by otfried** [comment]
It seems a bit heavy to add a dictionary with arbitrary mappings to each object - some people make drawings with thousands of objects (but to be honest, Ipe doesn't really handle very large documents so well anyway).

One simple idea for the moment would be to make each such object a one-element group, and store your attribute in the group's URL (that is essentially a free-form string attribute, as long as you don't export to PDF).

**Comment by thobl** [comment]
Adding a global dictionary might indeed be a bit heavy. However, I don't really need such a global dictionary but rather an attribute where I can store custom data. As long as this attribute is not used, this should not have any impact on the performance and there are probably several other attributes that are rarely used (e.g., the "url"-attribute for groups you mentioned).

The url-attribute would be a good attribute to store custom data, if I could apply it to arbitrary objects and not just groups. Using a dummy-group for every object makes things kind of messy (e.g., if the user wants to later edit the objects, they first have to ungroup them, which means that the url-attribute gets lost).

**Comment by otfried** [comment]
Well, users do make figures with 10'000 path objects, but not with 10'000 groups - that's why additional features like clipping paths, urls, decorations are on groups and not on every object.

You are aware that there is an "Edit group" function that allows you to edit the object without ungrouping.

Ipe attributes are typically stored in 4 Bytes, so they are not expensive.   But the way you asked, it seemed like you wanted to be able to do
```
obj:set("myCustomProperty1", "value1")
obj:set("myCustomProperty2", "value2")
obj:set("myCustomProperty3", "value3")
obj:set("myCustomProperty4", "value4")
```
and that would mean that I have to keep a dictionary of key-value pairs for the object (Of course it could be a pointer to a dictionary that is usually null, but this would take quite some extra work to make the undo-mechanism work right).

What I could offer in the short term is a single attribute called, say, "data" that is not used by Ipe but stored with the object.

**Comment by thobl** [comment]
Makes sense.

A single data-attribute that works for any type of object would be great, thank you!

**Comment by otfried** [comment]
Implemented in 7.2.25.
